component/newstorage/mem/mem.go

```diff
@@ -0,0 +1,372 @@
+	}
+}
+
+// Delete deletes the key + value pair (and all tags) associated with key.
+// If key is empty, then an error will be returned.
+func (m *memStore) Delete(k string) error {
+	if k == "" {
+		return errEmptyKey
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	delete(m.db, k)
+
+	return nil
+}
+
+// Batch performs multiple Put and/or Delete operations in order.
+// If any of the given keys are empty, then an error will be returned.
+func (m *memStore) Batch(operations []newstorage.Operation) error {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, operation := range operations {
+		if operation.Key == "" {
+			return errEmptyKey
+		}
+	}
+
+	for _, operation := range operations {
+		if operation.Value == nil {
+			delete(m.db, operation.Key)
+			continue
+		}
+
+		m.db[operation.Key] = dbEntry{
+			value: operation.Value,
+			tags:  operation.Tags,
+		}
+	}
+
+	return nil
+}
+
+// Close closes this store object, freeing resources and deleting all data stored within.
+func (m *memStore) Close() error {
+	m.Lock()
+	defer m.Unlock()
+
+	m.db = make(map[string]dbEntry)
+
+	m.close(m.name)
+
+	return nil
+}
+
+func (m *memStore) getMatchingKeysAndDBEntries(tagName, tagValue string) ([]string, []dbEntry) {
+	var matchAnyValue bool
+	if tagValue == "" {
+		matchAnyValue = true
+	}
+
+	var keys []string
+
+	var dbEntries []dbEntry
+
+	for key, dbEntry := range m.db {
+		for _, tag := range dbEntry.tags {
+			if tag.Name == tagName && (matchAnyValue || tag.Value == tagValue) {
+				keys = append(keys, key)
+				dbEntries = append(dbEntries, dbEntry)
+
+				break
+			}
+		}
+	}
+
+	return keys, dbEntries
+}
+
+// memIterator represents a snapshot of some set of entries in a memStore.
+type memIterator struct {
+	currentIndex   int
+	currentKey     string
+	currentDBEntry dbEntry
+	keys           []string
+	dbEntries      []dbEntry
+}
+
+// Next moves the pointer to the next entry in the iterator. It returns false if the iterator is exhausted.
+func (m *memIterator) Next() (bool, error) {
+	if len(m.dbEntries) == m.currentIndex || len(m.dbEntries) == 0 {
+		return false, nil
+	}
+
+	m.currentKey = m.keys[m.currentIndex]
+	m.currentDBEntry = m.dbEntries[m.currentIndex]
+	m.currentIndex++
+
+	return true, nil
+}
+
+// Key returns the key of the current entry.
+func (m *memIterator) Key() (string, error) {
+	if len(m.keys) == 0 {
+		return "", errIteratorExhausted
+	}
+
+	return m.currentKey, nil
+}
+
+// Value returns the value of the current entry.
+func (m *memIterator) Value() ([]byte, error) {
+	if len(m.keys) == 0 {
+		return nil, errIteratorExhausted
+	}
+
+	return m.currentDBEntry.value, nil
+}
+
+// Tags returns the tags associated with the key of the current entry.
+func (m *memIterator) Tags() ([]newstorage.Tag, error) {
+	if len(m.keys) == 0 {
+		return nil, errIteratorExhausted
+	}
+
+	return m.currentDBEntry.tags, nil
+}
+
+// Close closes this iterator object, freeing resources.
+func (m *memIterator) Close() error {
+	m.keys = make([]string, 0)
```

Comment:
Not sure if this is necessary. The garbage collector should take care of it.

Reply:
Removed the unnecessary lines.